### ISSUE TITLE
`@serializable` seems verbose to me

### ISSUE THREAD

**Issue by crazymonkyyy** [opendlang/d#7]
@GrimMaple 

Id have to test to be sure, but I assume your json code given a plain-old-struct would do nothing(and plain-old structs are all I care about)

id suggest https://github.com/opendlang/d/blob/0ec1d63a7112af7e136165e0566d944286f810b6/source/odc/serialization/package.d#L157

gets logic like "if its a struct and length of the filter is 0, assume all fields are serializable", and seeing if it just werks

**Comment by GrimMaple** [comment]
> @GrimMaple
> 
> Id have to test to be sure, but I assume your json code given a plain-old-struct would do nothing(and plain-old structs are all I care about)

Yes, you **have to** explicitly mark anything you want to serialize as `@serializable`. It's by design.

> id suggest
> 
> https://github.com/opendlang/d/blob/0ec1d63a7112af7e136165e0566d944286f810b6/source/odc/serialization/package.d#L157
> 
> gets logic like "if its a struct and length of the filter is 0, assume all fields are serializable", and seeing if it just werks

There have been a lot of talk about this, and it was a sort-of community agreement that this is a bad idea. Look at your `stack` thing for example. If my serializable code would be able to just serialize it, then you'd get a JSON like this:
```json
{
  "lastindex": 123,
  "length": 123,
  "opSlice": {
    "data": 12345678,
    "key": 1,
    "until": 123,
    // ... etc
  }
}
```

Serializable code now suggest that if you want to serialize a non-serializable, you'd do something like this:
```d
struct SerializableStack
{
  @serializable int[] stack() { return whatever; }
  private stack!int s;
}
```

This being said, I think it's possible to do something like this:
```d
@serializable struct MyThing
{

}
```
which would mark `MyThing` as "serialize everything". Though I think it's a bad idea on practice.

**Comment by adamdruppe** [comment]
The problem with the uda thing is sometimes you need to modify it from the outside - some lib thing that you can't even edit - and yet want the rest of it to remain automatic. I think C# has a way to go back up but id be thinking like a per-field filter that goes like type, name, uda present and returns some command to the serializer.

idk id have to try it

**Comment by crazymonkyyy** [comment]
> It's by design.
> it was a sort-of community agreement that this is a bad idea.

I thought so, but this is safetism that bugs me while I doubt I can convince you of my full rant, I think you should have an escape hatch at least

> @serializable struct MyThing

I believe this is a bad idea, id have to check your template logic but, I think full forwarding options is the best practice and not the uda things people try

> json "opSlice": {

hmmm Im confused why this is the case; if this is just the wrong traits do you oppose specail behavoir on struct in general or because of this happened when you tried?

**Comment by GrimMaple** [comment]
> I believe this is a bad idea, id have to check your template logic but, I think full forwarding options is the best practice and not the uda things people try

Full forwarding is not the best practice and literally none of the [sane] serializers do this. Even mentioned C# doesn't do it, it only forwards Properties (the get;set thing) by default. And even then, you have both `JsonInclude` attribute (to include non-properties) and `JsonExclude` (to exclude properties). 

The thing is, assuming that everything is serializable out of the box causes a lot of issues, while explicit thing doesn't cause issues __at all__. Well, besides mild inconvenience of having to mark it explicitly, which can be solved with `@serializable struct MyThing`.
 
> > json "opSlice": {
> 
> hmmm Im confused why this is the case; if this is just the wrong traits do you oppose specail behavoir on struct in general or because of this happened when you tried?

PODs are a very small subset of what happens when serialization is actually needed; most often you will have objects with at least _some_ logic in them. Those objects often have getters/setters to do some dynamic updating or other things. So, you have to also consider that when serializing. D has `@property`, but it sucks so much I'd suggest we remove it from the lang and make a `@deprecated struct property{}` instead.

**Comment by crazymonkyyy** [comment]
> Full forwarding is not the best practice and literally none of the [sane] serializers do this. 

Im not sure what you think I mean; I mean a template and api thing

when I make stack I have a size option, if something uses stack (and it wants to have the nogc option) it should simply forward the size argument

```d
struct usesstack(int size=-1){
  stack!(T,size) data;
  ...
```
with no futher effort, this will reuse stacks internal logic and coherence in api; very possibly making it nogc and fast; I can offer more complex examples but they maybe a bit harder to understand.

if you have a `serializetojson(bool force=false)` and you pass the argument along, you could only attempt to serialize something without udas with the user makes a silly worse call.

> PODs are a very small subset of what happens when serialization is actually needed

if it would just work Id use more serialization, its not nessery I can also just parse strings in the laziest way I think of after 10 seconds of thinking, more of a wishlist then a requirement but better io should just be on the todo list for making a better tool box; while I could shift 95% of my io use cases to cli's and printf debugging, piping, I shouldnt in lets say >35% of cases.

**Comment by GrimMaple** [comment]
> if you have a `serializetojson(bool force=false)` and you pass the argument along, you could only attempt to serialize something without udas with the user makes a slightly worse call.

It doesn't change anything at all; still same issues, just different sin tax. I don't want to allow "attempting" to serialize everything because it's error prone, unstable, and generally leads to hard-to-deal-with issues like recursion and versioning. And you'd still have to use UDAs because you'd have to selectively limit what can be serialized, so it doesn't really worth the struggle. Just design your objects properly and that would be it.

Realistically, I don't want to allow `@serializable struct MyThing` too because it leads to difficulties. Imagine if one day you add a member to `MyThing` -- your entire serialization can fail with no meaningful indication to it. Why do you want to serialize everything anyway? Why slapping `@serializable` is such a big problem?

> if it would just work Id use more serialization, its not nessery I can also just parse strings in the laziest way I think of after 10 seconds of thinking, more of a wishlist then a requirement but better io should just be on the todo list for making a better tool box; while I could shift 95% of my io use cases to cli's and printf debugging, piping, I shouldnt in lets say >35% of cases.

JSON isn't the only available option for serialization, and it's realistically not even the best one. Current solution provides an adequate across the board solution, not just JSON. It would be super easy to create an XML or YAML serializator with provided library code :)

**Comment by crazymonkyyy** [comment]
serialization is relatively low on my todo list so Im going to drop this until I have a usecase in front of me

but as written I would expect I would need to add a bunch of uda's to my data structures to get them to work

> Why slapping @serializable is such a big problem?

composite and voldemort types

`stack!(Tuple(Vector2,int),5)`

Is my simple tuple implantation suppose to detect if the members are serializable to aggregate up the uda for you? Or do you want to dig down, find *all* of my composite types(I could easily make 100 such cases once I start writing algorithms) recursively and figure it out?

or...  have a trust-the-programmer escape hatch and then I learn that one detail and move on

> Just design your objects properly

unlikely

---

old code link: https://github.com/crazymonkyyy/comfysettings would do this differently but still hold many of the same opinion I formed about serialization from this

**Comment by GrimMaple** [comment]
> Is my simple tuple implantation suppose to detect if the members are serializable to aggregate up the uda for you? Or do you want to dig down, find _all_ of my composite types(I could easily make 100 such cases once I start writing algorithms) recursively and figure it out?

Did you read what I wrote earlier? I provided an exhaustive argument on why I didn't design serialization around guessing what could be serialized and what couldn't. You provide basic examples, and even then it's unclear what the result should be: how do you expect such a `stack` to be serialized and why?

I designed serialization in such a way that it would be easy to understand and use, without complex rules and trickery needed to make it work. Attempting to auto-serialize unmarked objects inevitably leads to complex serialization rules (this can be serialized, this can't) and runtime issues. Serialization loops, versioning issues - this is what you get when you try to be too smart about it. Yet marking everything explicitly is error-safe, easy to understand and equally easy to use - you don't have to guess how your objects are serialized, you can know it 100% by just looking at the code. Nothing shady is gonna happen in the background.
